### DISCUSSION TITLE
Prevent selection state sync in multiplayer

### DISCUSSION THREAD

**Discussion by ysds** [xyflow/xyflow#2570] [General]
Hello dear folks,
I have been developing a collaborative flow for multiple users with React Flow and a CRDT lib, like a https://reactflow.dev/docs/examples/interaction/collaborative/.

For example, in a scenario where one user selects a node and deletes it, another user can steal the selection state, thus potentially breaking the typical editorial collaboration UX. To avoid this, I'd like to allow each user to store their selection state locally.

If you have any ideas how to solve this, please let me know.
Thanks :)

**Comment by hankportney** [comment]
@ysds I had this exact same problem, and solved it by making selection client-side, as you suggested.  Specifically:

1. create a local state `selectedNodes` on each client to track which nodes are selected
2. inside `changes.forEach`, intercept `change.type === "select"` node change events...
- adding those nodes to `selectedNodes` if `change.selected === true`
- removing those nodes from `selectedNodes` if `change.selected === false`
- exempting the change from being applied to the shared state by ***NOT*** running `sharedTypes.nodes.set(change.id, node)`;
3. inside `changes.forEach`, intercept `change.type === "remove"` node change events to remove the node from `selectedNodes` if it was selected
4. intercept the `setNodes` call inside the `sharedType.observe` observer function to apply `selected: true` to nodes in `selectedNodes`; see below code snippet.

 - `
setNodes(
        Array.from(sharedTypes.nodes.values()).map((node) =>
          selectedNodes.find((match) => match.id === node.id)
            ? { ...node, selected: true }
            : node
        )
      );
`

Additionally, while I haven't done this yet, **if I felt selection still needed to be shared without being "stolen"**, I would likely store each client's selection in the awareness protocol, and have that affect the render and selectability of nodes on each client.

Best of luck!
